csharp/src/Google.Protobuf/IExtendableMessage.cs

```diff
@@ -0,0 +1,78 @@
+﻿#region Copyright notice and license
+// Protocol Buffers - Google's data interchange format
+// Copyright 2015 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+using Google.Protobuf.Collections;
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Generic interface for a Protocol Buffers message containing one or more extensions, where the type parameter is expected to be the same type as the implementation class
+    /// </summary>
+    public interface IExtendableMessage<T> : IMessage<T> where T : IExtendableMessage<T>
+    {
+        /// <summary>
+        /// Gets the value of the specified extension
+        /// </summary>
+        TValue GetExtension<TValue>(Extension<T, TValue> extension);
+
+        /// <summary>
+        /// Gets the value of the specified repeated extension or null if the extension isn't registered in this set.
+        /// For a version of this method that never returns null, use <see cref="IExtendableMessage{T}.GetOrRegisterExtension{TValue}(RepeatedExtension{T, TValue})"/>
+        /// </summary>
+        RepeatedField<TValue> GetExtension<TValue>(RepeatedExtension<T, TValue> extension);
+
+        /// <summary>
+        /// Gets the value of the specified repeated extension, registering it if it isn't
+        /// </summary>
+        RepeatedField<TValue> GetOrRegisterExtension<TValue>(RepeatedExtension<T, TValue> extension);
```

Comment:
it seems odd that we have a GetOrRegisterExtension for RepeatedExtension but we don't have it for a non-repeated Extension. What is the reason?

Reply:
`GetExtension` will get the default value of the field if it isn't present in the set and doesn't register the field. Repeated fields don't have a default, so they return `null`. But `null` isn't normally expected for a `RepeatedField` value, so you could make it so that `GetExtension` for repeated fields returns non-`null`, but could cause allocations if the value isn't used. So to allow for the most control, `GetExtension` won't allocate and returns `null`, `GetOrRegisterExtension` will allocate and returns a normal value, and `ClearExtension` clears the field, making `GetExtension` return `null` again.

Reply:
It the term "register" in the meaning of "set the repeated extension to the default empty value" a name that you came up with or do other languages use similar terminology? My understanding of "register" is slightly different so this method's name seems a bit confusing to me. 

Idea:
How about adding an `RepeatedField<TValue> GetExtension<TValue>(RepeatedExtension<T, TValue> extension, bool initializeValue)` overload instead ?  `initializeIfNotPresent = false` would give you the GetExtension behavior and `initializeIfNotPresent = true` corresponds to GetOrRegisterExtension?

Reply:
Generally other languages use `Register` to mean they set up the extension to be tracked by either adding them to a registry or some other means. In C++ it adds them to the global registry, in Java you use it to add them to a registry. So I took it to mean that the extension was "tracked" and would exist in the set if you accessed it. But I guess since that definition isn't universal.

I'm open to renaming the method to something easier to understand from a brief glance. I don't think `GetExtension(RepeatedInt32Extension, true)` would be that easy to read from an outside perspective. Perhaps something like `Insert` or `Initialize` or some other word over `Register`.

Reply:
@ObsidianMinor  thanks for the explanation, but I think that transition from the original meaning of "Register" in other languages to the way you used it should be avoided to maintain consistency. In the context of this PR, the "Register" also sounds like it'd have to do with ExtensionRegistry ("registering the fact that an extension was declared for a message") while in fact it doesn't (it has to do with whether the value of an extension was set or not).

I think `GetOrInitializeExtension` is a definitely better than GetOrRegisterExtension, but it's still not perfect IMHO